### PR TITLE
Add option to skip deepcopy on dag_to_circuit

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds a new option, copy_operations, which allows to disable the deepcopy on dag_to_circuit(). Previously the dag to circuit converter would always deep copy the operations, however in the transpiler we don't need this extra overhead because the DAGCircuits are temporary internal artifacts and will not be resused outside of the transpiler. So we can avoid the copy overhead in these situations, however in general the converter needs to default to copying. This new argument enables this workflow by keeping the current behavior by default but enabling the transpiler to skip copying operations in the output conversion.

### Details and comments

### CODE REVIEW

**Comment by Cryoris** [comment]
It might be nice to have the same for `circuit_to_dag`, that would e.g. allow to skip the deepcopy upon `QuantumCircuit.__eq__`.

**Comment by Cryoris** [comment]
On a first look there are some other places we we probably could skip the deepcopy:
* `quantumcircuit.py`, L4811 in `_circuit_from_qasm`: the DAG is newly constructed inside the `ast_to_dag` and doesn't need to be copied again 
* `basis_translator.py` inside the logs of `_compose_transforms`: the DAG is only logged so doesn't need to be copied

Plus some places where the dag was previously copied in `circuit_to_dag` and doesn't need to be copied a second time (like the cases you fixed already in the PR):
* `transpiler.py`, L562
* `DAGCircuit.reverse_ops`
* `RunningPassManager.run`
* `control_flow.py` inside the function `map_blocks`

Should we update those in this PR, too?

File: qiskit/converters/dag_to_circuit.py [hunk 1]
```diff
@@ -60,7 +67,10 @@ def dag_to_circuit(dag):
     circuit.calibrations = dag.calibrations
 
     for node in dag.topological_op_nodes():
-        circuit._append(CircuitInstruction(copy.deepcopy(node.op), node.qargs, node.cargs))
+        op = node.op
+        if copy_operations:
+            op = copy.deepcopy(op)
+        circuit._append(CircuitInstruction(op, node.qargs, node.cargs))
```
Comment (jakelishman) [inline qiskit/converters/dag_to_circuit.py:73]: We don't have an optimising compiler to help us here, so it's possible it _might_ be a measurable improvement to move the `if copy_operations` outside the loop, especially in the case where it's false.

File: qiskit/converters/dag_to_circuit.py [hunk 2]
```diff
@@ -16,11 +16,18 @@
...
 
 
-def dag_to_circuit(dag):
+def dag_to_circuit(dag, copy_operations=True):
     """Build a ``QuantumCircuit`` object from a ``DAGCircuit``.
 
     Args:
         dag (DAGCircuit): the input dag.
+        copy_operations (bool): Deep copy the operation objects
+            in the :class:`~DAGCircuit` for the output :class:`~.QuantumCircuit`.
```
Comment (jakelishman) [inline qiskit/converters/dag_to_circuit.py:25]: You'll need a leading `.` on the `DAGCircuit` cross-ref too.

**Comment by mtreinish** [comment]
> On a first look there are some other places we we probably could skip the deepcopy:
> 
>     * `quantumcircuit.py`, L4811 in `_circuit_from_qasm`: the DAG is newly constructed inside the `ast_to_dag` and doesn't need to be copied again

I'll add this now.

> 
>     * `basis_translator.py` inside the logs of `_compose_transforms`: the DAG is only logged so doesn't need to be copied
> 

I'll add this too, but this is still super inefficient (which is why it has a debug level guard I guess. We really need to add a circuit drawer interface for `DAGCircuit` because the circuit drawers also are internally converting circuits to dags to get the topological ordering of operations

**Review by Cryoris — APPROVED** [review]
LGTM, thanks for the additions 👍🏻

**Pull request merged**